csrc/lower_thread_predicate.cpp

```diff
@@ -494,6 +504,221 @@ class RedundantUseAnalysis : BackwardVisitor {
+  // For each broadcast root domain, find a concretized domain from its
+  // exact mapped domain set.
+  void setConcretizedBroadcastRootDomain() {
+    for (auto rd : root_domain_) {
+      if (rd->isBroadcast()) {
+        const auto& mapped_rd_exact_set = GpuLower::current()
+                                              ->caMap()
+                                              ->getIdSets(IdMappingMode::EXACT)
+                                              .getDisjointSetOf(rd);
+        for (auto mapped_rd : mapped_rd_exact_set) {
+          std::unordered_set<IterDomain*> all_cids =
+              GpuLower::current()
+                  ->concretizedBroadcastDomains()
+                  ->allConcretizedDomains(mapped_rd);
+          auto fromSameExactGroup = [&all_cids]() {
+            const auto& exact_set = GpuLower::current()
+                                        ->caMap()
+                                        ->getIdSets(IdMappingMode::EXACT)
+                                        .getDisjointSetOf(*all_cids.begin());
+            return std::all_of(
+                all_cids.begin(), all_cids.end(), [&](IterDomain* id) {
+                  return exact_set.has(id);
+                });
+          };
+          if (!all_cids.empty()) {
+            // assert that all_cids are in the same group in the
+            // IdMappingMode::EXACT graph
+            TORCH_INTERNAL_ASSERT(
+                fromSameExactGroup(),
+                "All concretized domains of a broadcast root domain should be in the same group in the IdMappingMode::EXACT graph.");
+            concretized_broadcast_root_domains_[rd] = *all_cids.begin();
+            break;
+          }
+        }
+      }
+    }
+  }
+  // Find all the root domains that are merged to the leaf domain.
+  // e.g. Root: [I1,B2,B3] -> Leaf: [I1*B2*B3]
+  std::vector<IterDomain*> getRootDomainsMergedToLeaf(IterDomain* ld) {
+    std::vector<IterDomain*> merged_root_domains;
+    std::vector<int> index_root_domain;
+    std::vector<IterDomain*> intermediate_domains = root_domain_;
+    auto all_exp = DependencyCheck::getAllExprsBetween(
+        {root_domain_.begin(), root_domain_.end()}, {ld});
+    for (auto expr : all_exp) {
+      if (auto merge = dynamic_cast<Merge*>(expr)) {
+        auto outer_iter_im = std::find(
+            intermediate_domains.begin(),
+            intermediate_domains.end(),
+            merge->outer());
+        auto inner_iter_im = std::find(
+            intermediate_domains.begin(),
+            intermediate_domains.end(),
+            merge->inner());
+        TORCH_INTERNAL_ASSERT(
+            outer_iter_im != intermediate_domains.end(),
+            "Couldn't find ",
+            merge->outer());
+        TORCH_INTERNAL_ASSERT(
+            inner_iter_im != intermediate_domains.end(),
+            "Couldn't find ",
+            merge->inner());
+        int dist = std::distance(inner_iter_im, outer_iter_im);
+        if (std::abs(dist) != 1) {
```

Comment:
Is this due to how the index is computed in `getIndexWithoutBroadcast`?

Reply:
This was due to the old algorithm. But with the current new method which directly using the index of the leaf domain calculated from the merged root domains, this restriction is no longer needed. I deleted and added a test case where the merged root domains are not neighbors. e.g.
```
  at::Tensor t0 = at::randn({8, 10, 12}, options);
  at::Tensor t1 = at::randn({8, 7, 10, 12, 9}, options);
  t0 --> {8, B7, 10, 12, B9} --> {8*B9, B7, 10, 12} --> {8*B9(BIDX), B7, 10, 12}
```
The generated kernel will use BIDX = 8*B9 = 72 blocks and the write condition is:
 `blockIdx.x ==  T1.size[4] * ( blockIdx.x / T1.size[4])`, e.g. blockIdx.x == 0, 9, 18, ...63